pkg/controller/cincinnati/cincinnati_controller.go

```diff
@@ -214,40 +232,117 @@ func handleCACertStatus(reqLogger logr.Logger, status *cv1beta1.CincinnatiStatus
 	reqLogger.Info(message)
 }
 
-func (r *ReconcileCincinnati) ensureAdditionalTrustedCA(ctx context.Context, reqLogger logr.Logger, instance *cv1beta1.Cincinnati, resources *kubeResources) error {
+func (r *ReconcileCincinnati) postAddPullSecret(ctx context.Context, reqLogger logr.Logger, instance *cv1beta1.Cincinnati, resources *kubeResources) error {
+	sourcePS, err := r.findPullSecret(ctx, reqLogger, instance, resources)
+	if err != nil {
+	   return err
+	} else if sourcePS == nil {
+	   return nil
+	}
+
+	resources.newPullSecret(instance, sourcePS)
+	resources.addPullSecret(instance)
```

Comment:
I'm hoping we can avoid modifying `resources` after instantiating it. What I'm proposing is that `ensurePullSecret` runs before `newKubeResources`. That way, when all the resources are being generated, it will already be able to consume the pull secret copy. WDYT? Does that make sense?

We would apply the same pattern to `ensureAdditionalTrustedCA`, even though I know that's not part of this PR.

Reply:
> I'm hoping we can avoid modifying resources after instantiating it. What I'm proposing is that ensurePullSecret runs before newKubeResources. That way, when all the resources are being generated, it will already be able to consume the pull secret copy. WDYT? Does that make sense?
>
Ya I understand.

The challenge with doing everything before `newKubeResources` though is we need a deployment object that the `ensurePullSecret` function can modify.  A lot of what I came up with I found to be outside of what the `new.go` functions do.  Maybe you have some ideas.   

What I like about the current change is that even though I'm doing some additional resource regeneration, I'm doing it out in the open.  So I consider the kubeResources loop and post modify loop combine to make up the "resource generation loops", which is much more clear.

At the end of the day, we're talking about generating an object a few additional times.  So I guess I consider the code change to me mostly cosmetic.  Maybe I missing something.

Reply:
Rather than modifying the Deployment, why not just have the deployment always reference the secret from the beginning? I assume it would go like this:

1) ensure things exist that are based on non-CR inputs: `ensurePullSecret` and `ensureAdditionalTrustedCA`.
2) generate resource definitions assuming the above has been accomplished. The resources created in step 1 will be referenced by these definitions.
3) ensure the generated resources exist, include all other `ensure*` functions.

Reply:
If I'm understanding you correctly, the deployment spec would look have this in it:

```yaml
                                                corev1.Volume{
                                                        Name: "cincinnati-graph-data",
                                                        VolumeSource: corev1.VolumeSource{
                                                                EmptyDir: &corev1.EmptyDirVolumeSource{},
                                                        },                                                                                                                                                                                                                                      
```

Reply:
If I'm understanding you correctly, the deployment spec would have these by default:

```yaml
...

                                Spec: corev1.PodSpec{                                                                                                                                                                                                                                
                                         Volumes: []corev1.Volume{ 
                                             VolumeSource: corev1.VolumeSource{                                                                                                                                                                                                                              
                                                Secret: &corev1.SecretVolumeSource{                                                                                                                                                                                                                  
                                                    SecretName: NamePullSecret,                                                                                                                                                                                                                  
                                                    DefaultMode: &mode,                                                                                                                                                                                                                  
                                                },                                                                                                                                                                                                                                   
                                              },                                                                  
                                              VolumeSource: corev1.VolumeSource{                                                                                                                                                                                                                                              
                                                 ConfigMap: &corev1.ConfigMapVolumeSource{                                                                                                                                                                                                            
                                                     DefaultMode: &mode,                                                                                                                                                                                                                          
                                                     LocalObjectReference: corev1.LocalObjectReference{                                                                                                                                                                                                              
                                                         Name: nameAdditionalTrustedCA(instance),                                                                                                                                                                                            
                                                     },                                                                                                                                                                                                                                           
                                                     Items: []corev1.KeyToPath{                                                                                                                                                                                                                           
                                                        corev1.KeyToPath{                                                                                                                                                                                                                                    
                                                        Path: "tls-ca-bundle.pem",
                                                        Key:  NameCertConfigMapKey,                                                                                                                                                                                                          
                                                     },                                                                                                                                                                                                                                   
                                                 },                                                                                                                                                                                                                                   
                                              },                                                                                                                                                                                                                                   
                                          },        
                                      },                                                                                                                                                                                                     
```

`ensurePullSecret` runs to setup the pull-secret copy.  `ensureAdditionalTrustedCA` runs to setup the externalCA configmap copy.  Normal resource generation occurs and the remaining `ensure*` functions run.

I think this will work for the pull-secret because we should always have a copy.   However, it's fine to [not have a externalCA configmap](https://github.com/openshift/cincinnati-operator/blob/master/pkg/controller/cincinnati/cincinnati_controller.go#L233), so I don't think the same pattern works there.

Reply:
Can we just pass the configmap's name as an optional parameter to `newKubeResources`? Given the choice of telling it up-front that the configmap exists, or waiting until later to add it, I think it's preferable to do it up-front.

Reply:
Following some offline chat, I think this is ok as-is and don't see a need to block the PR.

My TL;DR: I think that `newKubeResources` is a constructor, and we should let it do all of the constructing, rather than defer some of the constructing to outside logic. This code as-is does make it clear where the deferred constructing should happen, which is helpful toward alleviating my concerns about leaking constructor logic outside of one well-known place. In this case, the logic is basically in two well-known places instead of one.